src/basic/rlimit-util.c

```diff
@@ -13,6 +13,10 @@
 #include "string-table.h"
 #include "time-util.h"
 
+static inline int rlimit_to_nice(rlim_t prio) {
+        return (PRIO_MAX - prio + 1);
```

Comment:
the () around this are unnecessary, please drop...

Reply:
So i just realized that the kernel does not restrict the value range for RLIMIT_NICE even though only a small subrange of it makes sense. This means the function above will result bogus values if people use the full rlim_t range (which is usually 64bit) and we return an int (i.e. 32bit). This function should hence be tweaked to saturate the results, and never return values outside the valid nice level range... THis is a biggie.

(I am just realizing this now, we might have similar issues elsewhere in our codebase)

Reply:
Why don't I actually use `nice()` here then? It kind of has the right semantics.

Reply:
nah, nice() in glibc doesn't do everything we want.

Reply:
I think we are fine w.r.t. using bogus values:

> Attempts to set a priority outside this range are silently clamped to the range.

Reply:
well, but this creates ambiguities. i.e. if i pass a really high rlim_t value here, then the top bits are silently dropped during the cast, and that's what I really don't like. We should catch these values before silently casting them. I mean, this is what we do pretty much everywhere: either catch the outlying values before casting and generate an error, or catch them and "saturate" them, i.e. converting them implicitly to the closest sensible value. In this case the letter approach sounds OK.

Let me give you an example. user sets the rlim for RLIMIT_NICE to 2^32-1. This function would hence calculate: `PRIO_MAX-2^32-1+1`, i.e. `20-2^32`, i.e. -4294967276, which then cast to `int` is 20, i.e. only allows really high nice values, even though the limit is set so high, so that one should expect it would alllow really low nice values, too... i.e. the user specified a limit the kernel happily accepted that means one thing, and we intrepret it due to the implict casting to mean the complete opposite. And that's bad!

hence, the function really should be something like this:

```
static int rlimit_to_nice(rlim_t prio) {
    if (prio <= 1)
            return PRIO_MAX-1; /* i.e. 19 */
    if (prio >= -PRIO_MIN + PRIO_MAX) 
            return PRIO_MIN; /* i.e. -20 */
    return PRIO_MAX - (int) prio;
}
```

(and probably should have a test associated, which tests a couple of inputs against the expected outputs)

with the above definition no implicit casts take place, and overly small rlim_t values will always saturate out at 19, and overlay large ones at -20, and there's the strict guarantee that this function only returns valie nice values.